usr/src/lib/libzfs/common/libzfs_fru.c

```diff
@@ -186,6 +190,75 @@ libzfs_fru_gather(topo_hdl_t *thp, tnode_t *tn, void *arg)
 }
 
 /*
+ * Given a disk FRU, check that FRU contains a slot number and remove FRU
+ * details that aren't needed when comparing FRUs by slot number.
+ */
+static char *
+diskfru_to_slot(libzfs_handle_t *hdl, const char *diskfru)
+{
+	nvlist_t *nvl, **hc;
+	char *hc_name, *tmp = NULL;
+	int ret, i;
+	uint_t hc_cnt;
+
+	/* string -> nvlist */
+	if (_topo_fmri_str2nvl(hdl->libzfs_topo_hdl, diskfru, &nvl, &ret) != 0)
+		return (NULL);
+
+	/* Need slot (bay) number in the FRU */
+	if (nvlist_lookup_nvlist_array(nvl, FM_FMRI_HC_LIST, &hc,
+	    &hc_cnt) != 0)
+		goto out;
+
+	for (i = 0; i < hc_cnt; i++) {
+		if (nvlist_lookup_string(hc[i], FM_FMRI_HC_NAME,
+		    &hc_name) == 0 && strcmp(hc_name, BAY) == 0)
+			break;
+	}
+	if (i == hc_cnt)
+		goto out;
+
+	/* Drop the unwanted components */
+	(void) nvlist_remove_all(nvl, FM_FMRI_HC_SERIAL_ID);
+	(void) nvlist_remove_all(nvl, FM_FMRI_HC_PART);
+	(void) nvlist_remove_all(nvl, FM_FMRI_HC_REVISION);
+
+	/* nvlist -> string */
+	if (_topo_fmri_nvl2str(hdl->libzfs_topo_hdl, nvl, &tmp, &ret) != 0)
+		tmp = NULL;
+out:
+	nvlist_free(nvl);
+	return (tmp);
+}
+
+/*
+ * Check if given FRUs match by slot number to skip comparing disk specific
+ * fields of the FRU.
+ */
+/* ARGSUSED */
+int
+libzfs_fru_cmp_slot(libzfs_handle_t *hdl, const char *a, const char *b,
+    size_t len)
+{
+	char *slota, *slotb;
+	int ret = -1;
+
+	if (a == NULL || b == NULL)
```

Comment:
I believe means that auto replace won't take place on drives, not in an enclosure or drives inside enclosures that don't report a bay number.
It would be nice to allow autoreplace on non-enclosure devices but that comes with the risk of screwing up jbod affinity in other use cases so I'm not sure there is a better way to deal with theses cases...